### ISSUE TITLE
Discrepancy between R's Keras and Python's Keras

### ISSUE THREAD

**Issue by CrisBMoya** [rstudio/keras#956]
Hi,

I been playing with the same data using Keras on both Python and R. I wrote a post on [StackOverflow explaining the situation](https://stackoverflow.com/questions/59462354/discrepancy-between-rs-keras-and-pythons-keras-accuracy-bug) but basically I'm having absolutely different Accuracy values when using Keras in R vs Keras on Python.

I invoke Keras by using `use_python()` option.
Here's my sessioninfo:

```
> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18362)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] keras_2.2.5.0    plyr_1.8.4       forcats_0.4.0    stringr_1.4.0    dplyr_0.8.3     
 [6] purrr_0.3.3.9000 readr_1.3.1      tidyr_0.8.3      tibble_2.1.3     ggplot2_3.2.1   
[11] tidyverse_1.2.1 

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.3           cellranger_1.1.0     pillar_1.4.2         compiler_3.6.1      
 [5] base64enc_0.1-3      tools_3.6.1          zeallot_0.1.0        jsonlite_1.6        
 [9] lubridate_1.7.4      gtable_0.3.0         nlme_3.1-140         lattice_0.20-38     
[13] pkgconfig_2.0.2      rlang_0.4.2          Matrix_1.2-17        cli_1.1.0           
[17] rstudioapi_0.10      haven_2.1.1          withr_2.1.2          xml2_1.2.2          
[21] httr_1.4.1           rappdirs_0.3.1       generics_0.0.2       vctrs_0.2.0         
[25] hms_0.5.0            grid_3.6.1           tidyselect_0.2.5     reticulate_1.14-9000
[29] glue_1.3.1           R6_2.4.1             readxl_1.3.1         whisker_0.4         
[33] modelr_0.1.5         magrittr_1.5         tfruns_1.4           backports_1.1.4     
[37] scales_1.0.0         rvest_0.3.4          assertthat_0.2.1     colorspace_1.4-1    
[41] tensorflow_2.0.0     stringi_1.4.3        lazyeval_0.2.2       munsell_0.5.0       
[45] broom_0.5.2 
```

**Comment by skeydan** [comment]
Can you provide a minimal reproducible example?

**Comment by CrisBMoya** [comment]
Yes,

Here´s a minimal reproducible example with the Bike Sharing dataset from Kaggle using R:
```
#Library
library(tidyverse)
library(keras)
library(RCurl)

#Use python
use_python("C:/Program Files/Python37/")

#Read data
KaggleBike="https://raw.githubusercontent.com/CrisBMoya/BikeSharingDataset/master/Data/train.csv"
TrainDF=read_delim(file=getURL(KaggleBike), delim=",")

#Subset
Y_Train=as.matrix(TrainDF[2001:nrow(TrainDF),c('count')])
X_Train=TrainDF[2001:nrow(TrainDF),c('temp', 'atemp', 'humidity', 'windspeed', 'casual', 'registered')]

#Data reshape
YVal=array_reshape(x=as.matrix(Y_Train), dim=c(8886, 1))
XVal=array_reshape(x=as.matrix(X_Train), dim=c(8886,2,3,1))

#Keras Model
model=keras_model_sequential() 
model %>% 
  layer_conv_2d(filters=10, kernel_size=c(2,2), padding='same', activation='relu') %>%
  layer_conv_2d(filters=15, kernel_size=c(2,2), padding='same', activation='relu') %>%
  layer_conv_2d(filters=20, kernel_size=c(3,3), padding='same') %>%
  layer_max_pooling_2d(pool_size=c(2,2), strides=1) %>%
  layer_flatten() %>%
  layer_dense(units=30, activation='relu') %>%
  layer_dense(units=20, activation='relu') %>%
  layer_dense(units=10, activation='relu') %>%
  layer_dense(units=1)

#Compile model
model %>% compile(
  loss = 'mse',
  optimizer = optimizer_adam(),
  metrics = c('accuracy'))

#Train model
model %>% fit(XVal, YVal, epochs = 30)
``

And the same thing but using Python:
```
import keras
import numpy as np
import pandas as pd

#Read dta
KaggleBike="https://raw.githubusercontent.com/CrisBMoya/BikeSharingDataset/master/Data/train.csv"
TrainDF=pd.read_csv(KaggleBike, parse_dates=["datetime"])

#Subset
X_Train=TrainDF[2000:]

#Subset TrainDF
YVal=np.array(X_Train[["count"]])
XVal=np.array(X_Train[["temp", "atemp", "humidity", "windspeed","casual","registered"]])

#Data reshape
YVal=YVal.reshape((8886,1))
XVal=XVal.reshape((8886,2,3,1))

#Create Keras 2D CNN Model
model=keras.Sequential()
model.add(keras.layers.Conv2D(filters=10, kernel_size=[2,2], padding='same', activation='relu'))
model.add(keras.layers.Conv2D(filters=15, kernel_size=[2,2], padding='same', activation='relu'))
model.add(keras.layers.Conv2D(filters=20, kernel_size=[3,3], padding='same'))
model.add(keras.layers.MaxPooling2D(pool_size=[2,2], strides=1))
model.add(keras.layers.Flatten())
model.add(keras.layers.Dense(units=30, activation='relu'))
model.add(keras.layers.Dense(units=20, activation='relu'))
model.add(keras.layers.Dense(units=10, activation='relu'))
model.add(keras.layers.Dense(units=1))

#Compile Model
model.compile(optimizer='adam', loss='mse', metrics=['accuracy'])

#Fit model to Data
model.fit(XVal, YVal, epochs=30)
```

**Comment by skeydan** [comment]
For that example, I get the same final accuracy `0.0047` from both R and Python - provided that I change your 

```
import keras
```

to 

```
import tensorflow.keras as keras
```

Without doing that, there is a big difference between both pieces of code: The R version uses Tensorflow Keras (www.tensorflow.org), while the Python one uses standalone Keras (http://keras.io).

For some background on this (admittedly, confusing) thing, see https://blogs.rstudio.com/tensorflow/posts/2019-10-08-tf2-whatchanges/.

**Comment by CrisBMoya** [comment]
Woah, well, thanks. I'll be reading that documentation to understand the difference in the accuracy as both codes ran well, but what it seems to be a _detail_ changes the outcome completely.

Thanks!
